### ISSUE TITLE
SQL Server Service Broker throws errors if this feature is not enabled for the database

### ISSUE THREAD

**Issue by BenedekFarkas** [OrchardCMS/Orchard#5600]
The "[SQL Server Service Broker](https://msdn.microsoft.com/en-us/library/bb522893.aspx)" feature utilizes a feature of SQL Server with the same name to be able to send async messages between databases in an efficient and reliable way. The problem seems to be that this feature is not enabled for databases by default (at least in SQL Server V12, which is the current default for SQL Azure), but it can be manually: either on a per-database level with a one-line query, or it can be enabled for all databases through the master DB.

The one-line query for enabling it on a DB is the following (the name of the DB can be grabbed from *ShellSettings.DataConnectionString* for example):
```
ALTER DATABASE myDatabase SET ENABLE_BROKER;
```
But: this query turns out to be running for a very long for time, because this command needs exclusive access to the DB, which is a very rare case ([see this article](http://rusanu.com/2006/01/30/how-long-should-i-expect-alter-databse-set-enable_broker-to-run)).

The workaround is to rollback all transactions and execute this query this way:
```
ALTER DATABASE myDatabase SET ENABLE_BROKER WITH ROLLBACK IMMEDIATE;
```
So my proposal for fixing this issue is:
- Add a migration class for the *Orchard.MessageBus.SqlServerServiceBroker* feature.
- In the *Create* method run the query to enable the Service Broker after creating a new transaction.
- In the *Uninstall* method run the query to disable the Service Broker (same query, but with *DISABLE_BROKER*).

**Comment by sebastienros** [comment]
You can use SQL broker without using Sql Server, or using a different instance, hence using a migration might be wrong.

**Comment by BenedekFarkas** [comment]
Right. Well, it's easy to determine by checking the ShellSettings if the DB is provided by SQL Server and then you definitely need to enable it (at least in the latest versions, but I've seen discussions about older versions of SQL Server regarding the same issue).
